File: web/packages/api-client/src/console/application-runtime/trajectory.ts
import { apiFetch } from '../../transport';

export type ProviderTrajectoryView = 'semantic' | 'protocol';

export interface ProviderTrajectoryOptions {
  request_id?: string;
  focus_event_id?: string;
}
export interface ProviderTrajectoryStep {
  links: Array<{
    relation: 'trigger' | 'context';
    flow_run_id: string;
    request_id: string;
    response_id?: string | null;
  }>;
  event_id: string;
  event_sequence: number;
  event_type: 'provider_semantic_step';
  created_at: string;
  metadata: {
    purpose: 'prewarm' | 'generate' | 'tool_resume' | 'compact' | 'unknown';
    run_mode?: string;
    duration_ms?: number | null;
    source: 'ai_native' | 'supplier_protocol';
    protocol?: string;
    transport?: 'http' | 'sse' | 'websocket';
    direction?: 'prepared' | 'received';
    kind:
      | 'model_call'
      | 'model_reply'
      | 'tool_call'
      | 'tool_result'
      | 'error'
      | 'observation_gap';
    status: 'recorded' | 'incomplete' | 'unavailable';
    step_key: string;
    preview?: string;
    tool_call_id?: string;
    provenance?: 'submitted_tool_result' | 'supplier_protocol' | 'ai_native';
    reason?: string;
    flow_run_id: string;
    node_id: string;
    node_run_id: string;
    invocation_id: string;
    provider_attempt_index: number;
    raw_sequence_start?: number;
    raw_sequence_end?: number;
  };
}
export interface ProviderTrajectoryPage {
  items: ProviderTrajectoryStep[];
  next_cursor: number | null;
  observation_count: number;
  persist_failed_count: number;
  protocol_integrity: 'complete' | 'incomplete' | 'pending' | 'not_recorded';
  protocol_persist_failed_count: number;
  integrity:
    | 'complete'
    | 'incomplete'
    | 'pending'
    | 'unavailable'
    | 'not_recorded';
}
export interface ProviderTrajectoryBody {
  sections: Array<{
    kind: 'configuration' | 'system' | 'context' | 'tools' | 'output' | 'error';
    value: unknown;
  }>;
  source: 'ai_native' | 'supplier_protocol';
  evidence_scope: 'step' | 'invocation';
  event_id: string;
  items: Array<{
    event_id: string;
    sequence: number;
    body: string;
    encoding: 'utf8' | 'base64';
  }>;
  next_cursor: number | null;
}
export function getConsoleProviderTrajectory(
  applicationId: string,
  runId: string,
  nodeRunId: string,
  cursor?: number,
  baseUrl?: string,
  options?: ProviderTrajectoryOptions
) {
  const query = new URLSearchParams({ limit: '50' });
  if (options?.request_id) query.set('request_id', options.request_id);
  if (options?.focus_event_id && cursor === undefined)
    query.set('focus_event_id', options.focus_event_id);
  if (cursor !== undefined) query.set('cursor', String(cursor));
  return apiFetch<ProviderTrajectoryPage>({
    path: `/api/console/applications/${applicationId}/logs/runs/${runId}/nodes/${nodeRunId}/trajectory?${query}`,
    baseUrl
  });
}
export function getConsoleProviderTrajectoryBody(
  applicationId: string,
  runId: string,
  nodeRunId: string,
  eventId: string,
  cursor?: number,
  view: ProviderTrajectoryView = 'semantic',
  baseUrl?: string
) {
  const query = new URLSearchParams({ limit: '8', view });
  if (cursor !== undefined) query.set('cursor', String(cursor));
  return apiFetch<ProviderTrajectoryBody>({
    path: `/api/console/applications/${applicationId}/logs/runs/${runId}/nodes/${nodeRunId}/trajectory/${eventId}?${query}`,
    baseUrl
  });
}

export function getConsoleRunTrajectory(
  applicationId: string,
  runId: string,
  cursor?: number,
  baseUrl?: string,
  options?: ProviderTrajectoryOptions
) {
  const query = new URLSearchParams({ limit: '50' });
  if (options?.request_id) query.set('request_id', options.request_id);
  if (options?.focus_event_id && cursor === undefined)
    query.set('focus_event_id', options.focus_event_id);
  if (cursor !== undefined) query.set('cursor', String(cursor));
  return apiFetch<ProviderTrajectoryPage>({
    path: `/api/console/applications/${applicationId}/logs/runs/${runId}/trajectory?${query}`,
    baseUrl
  });
}

export function getConsoleRunPayload(
  applicationId: string,
  runId: string,
  section: 'input_payload' | 'output_payload',
  baseUrl?: string
) {
  return apiFetch<Record<string, unknown>>({
    path: `/api/console/applications/${applicationId}/logs/runs/${runId}/payloads/${section}`,
    baseUrl
  });
}

File: api/crates/control-plane-contracts/src/ports/runtime/trajectory.rs
use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Copy, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ProviderTrajectoryView {
    #[default]
    Semantic,
    Protocol,
}

#[derive(Debug, Clone, Serialize)]
pub struct ProviderTrajectoryStep {
    pub event_id: Uuid,
    pub event_sequence: i64,
    pub event_type: String,
    pub created_at: String,
    /// Narrow, persisted metadata; never contains protocol body.
    pub metadata: serde_json::Value,
    pub links: Vec<WorkflowTrajectoryLink>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowTrajectoryLink {
    pub relation: String,
    pub flow_run_id: Uuid,
    pub request_id: Uuid,
    pub response_id: Option<String>,
}

#[derive(Debug)]
pub struct TrajectoryTargetNotFound;
impl std::fmt::Display for TrajectoryTargetNotFound {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("trajectory target not found")
    }
}
impl std::error::Error for TrajectoryTargetNotFound {}

/// Internal index selection; target_id is the native event or client step identity.
#[derive(Debug, Default, Clone, Copy)]
pub struct TrajectorySelection {
    pub request_id: Option<Uuid>,
    pub target_id: Option<Uuid>,
}

#[derive(Debug, Clone, Serialize)]
pub struct WorkflowEventSection {
    pub kind: String,
    pub value: serde_json::Value,
}

#[derive(Debug, Clone, Serialize)]
pub struct ProviderTrajectoryPage {
    pub items: Vec<ProviderTrajectoryStep>,
    pub next_cursor: Option<i64>,
    pub observation_count: i64,
    pub persist_failed_count: i64,
    pub integrity: String,
    pub protocol_integrity: String,
    pub protocol_persist_failed_count: i64,
}

#[derive(Debug, Clone, Serialize)]
pub struct ProviderTrajectoryEvidence {
    pub event_id: Uuid,
    pub sequence: i64,
    pub body: String,
    pub encoding: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct ProviderTrajectoryBody {
    pub event_id: Uuid,
    pub source: String,
    pub evidence_scope: String,
    pub sections: Vec<WorkflowEventSection>,
    pub items: Vec<ProviderTrajectoryEvidence>,
    pub next_cursor: Option<i64>,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ApplicationRunPayloadSection {
    InputPayload,
    OutputPayload,
}

#[async_trait]
pub trait ProviderTrajectoryRepository: Send + Sync {
    async fn provider_trajectory_filtered_page(
        &self,
        flow_run_id: Uuid,
        node_run_id: Option<Uuid>,
        cursor: Option<i64>,
        limit: i64,
        selection: TrajectorySelection,
    ) -> anyhow::Result<ProviderTrajectoryPage>;
    async fn client_trajectory_filtered_page(
        &self,
        flow_run_id: Uuid,
        node_run_id: Option<Uuid>,
        cursor: Option<i64>,
        limit: i64,
        selection: TrajectorySelection,
    ) -> anyhow::Result<super::ClientTrajectoryPage>;

    async fn provider_run_trajectory_page(
        &self,
        flow_run_id: Uuid,
        cursor: Option<i64>,
        limit: i64,
    ) -> anyhow::Result<ProviderTrajectoryPage>;
    async fn application_run_payload(
        &self,
        application_id: Uuid,
        flow_run_id: Uuid,
        section: ApplicationRunPayloadSection,
    ) -> anyhow::Result<Option<serde_json::Value>>;
    async fn provider_trajectory_page(
        &self,
        flow_run_id: Uuid,
        node_run_id: Uuid,
        cursor: Option<i64>,
        limit: i64,
    ) -> anyhow::Result<ProviderTrajectoryPage>;
    async fn provider_trajectory_body(
        &self,
        flow_run_id: Uuid,
        node_run_id: Uuid,
        event_id: Uuid,
        cursor: Option<i64>,
        limit: i64,
        view: ProviderTrajectoryView,
    ) -> anyhow::Result<Option<ProviderTrajectoryBody>>;
}
